### PR TITLE
signal-desktop: replace unlicensed Apple emoji

### PR DESCRIPTION
**Pull request by emilazy**
## Description of changes

Signal ships the Apple emoji set without a licence via an npm package and upstream does not seem terribly interested in fixing this; see:

* <https://github.com/signalapp/Signal-Android/issues/5862>
* <https://whispersystems.discoursehosting.net/t/signal-is-likely-violating-apple-license-terms-by-using-apple-emoji-in-the-sticker-creator-and-android-and-desktop-apps/52883>

I don’t want to mark Signal as `lib.licenses.unfree`, so this change instead replaces the bundled Apple emoji PNGs with ones generated from our freely‐licensed Noto Color Emoji font.

I chose Noto Color Emoji because it is the best‐maintained FOSS emoji font, and because Signal Android will also use the Noto emoji if the “Chats → Keyboard → Use system emoji” setting is turned on. Therefore, Noto Color Emoji is both within the bounds of the Signal user experience on other platforms, and more likely to match the emoji font installed on a NixOS system to boot. I have verified that Noto Color Emoji covers all the standalone emoji that the bundled Apple set does, and could not find any emoji sequence that reliably displayed correctly in Signal before these changes but did not afterwards. (Though I sure did find a good number of emoji that displayed weirdly in Signal both before and after.)

Signal will also download and cache large versions of the Apple emoji from their own update server at runtime. This does not pose a copyright concern for the Nixpkgs cache, but would result in inconsistent presentation between small and large emoji. Therefore, we also point these to our Noto Color Emoji PNGs, and gain a little privacy in the process.

**No invasive patches are made to the Signal code;** the only changes are to replace the unlicensed Apple emoji files with our own, and replace the URL that large versions are fetched from to point to them. There is no functional change to the application other than showing different images on the client and not requesting the jumbomoji pack files from the Signal update server. Ideally we’d build this package from source and simply omit the problematic files in the first place, but apparently that’s a little tricky and we should solve the compliance problem now.

The best solution, of course, would be for Signal to replace their unlicensed copy of Apple’s emoji with a freely‐licensed set compatible with their AGPLv3 licence. I may try and raise this situation again with Signal, although given the past response I am not optimistic, but I wanted to first address the potential copyright violation in Nixpkgs as swiftly as possible.

Although the Python script used to copy and rename the Noto PNGs is very simple, I have extensively documented it to help increase confidence in it and ease further maintenance. To reflect my willingness to keep this change maintained and take responsibility for it, I have added myself to the package maintainer list.

These changes actually result in the uncompressed size of the resulting package decreasing from 450 MiB to 435 MiB; as Signal would ordinarily download and cache up to 27 MiB of jumbomoji sheets from their servers during use, the effective disk space savings are likely to be higher.

Thanks to @mjm for helping test this.

I considered offering a `withUnfree` flag to allow the use of the bundled and downloaded Apple emoji, but it meaningfully complicates the derivation. If people feel very strongly about keeping the unlicensed Apple emoji, I can rework this change to add a flag at the cost of additional code complexity. (Although it’s arguably more `lib.licenses.illegal` than `lib.licenses.unfree`…)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by equirosa** [comment]
Result of `nixpkgs-review pr 337161` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>signal-desktop</li>
    <li>signal-desktop-beta</li>
  </ul>
</details>

**Review by equirosa — APPROVED** [review]

File: pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    True
+    """
+    return "_".join(
+        f"{ord(scalar_value):04x}"
+        for scalar_value in emoji
+        if scalar_value != "\ufe0f"
+    )
+
+
+def emoji_to_emoji_data_name(emoji):
```
Comment (Mic92) [inline pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py:58]:
```suggestion
def emoji_to_emoji_data_name(emoji: str) -> str:
```

File: pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    f"Missing Noto emoji: {emoji} {signal_emoji_name}",
+                    file=sys.stderr,
+                )
+                continue
+
+            (
+                emoji_data_out_path / f"{emoji_to_emoji_data_name(emoji)}.png"
+            ).symlink_to(out_path / emoji)
+
+    print(out_path.relative_to(asar_root))
```
Comment (Mic92) [inline pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py:115]: avoid introducing globals:

```suggestion
def main() -> None:
    noto_png_path, asar_root = sys.argv[1:]
    noto_png_path = Path(noto_png_path)
    asar_root = Path(asar_root).absolute()

    out_path = asar_root / "images" / "nixpkgs-emoji"
    out_path.mkdir(parents=True)

    emoji_data_out_path = (
        asar_root
        / "node_modules"
        / "emoji-datasource-apple"
        / "img"
        / "apple"
        / "64"
    )
    emoji_data_out_path.mkdir(parents=True)

    jumbomoji_json_path = asar_root / "build" / "jumbomoji.json"
    with jumbomoji_json_path.open() as jumbomoji_json_file:
        jumbomoji_packs = json.load(jumbomoji_json_file)

    for signal_emoji_names in jumbomoji_packs.values():
        for signal_emoji_name in signal_emoji_names:
            emoji = signal_name_to_emoji(signal_emoji_name)

            try:
                shutil.copy(
                    noto_png_path / f"emoji_u{emoji_to_noto_name(emoji)}.png",
                    out_path / emoji,
                )
            except FileNotFoundError:
                print(
                    f"Missing Noto emoji: {emoji} {signal_emoji_name}",
                    file=sys.stderr,
                )
                continue

            (
                emoji_data_out_path / f"{emoji_to_emoji_data_name(emoji)}.png"
            ).symlink_to(out_path / emoji)

    print(out_path.relative_to(asar_root))
  
 if __name__ == "__main__":
   main()
```

File: pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+directly to them.
+"""
+
+import json
+import shutil
+import sys
+from pathlib import Path
+
+
+def signal_name_to_emoji(signal_emoji_name):
```
Comment (Mic92) [inline pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py:21]:
```suggestion
def signal_name_to_emoji(signal_emoji_name: str) -> str:
```

File: pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    True
+    """
+    hex_bytes = zip(signal_emoji_name[::2], signal_emoji_name[1::2])
+    emoji_utf_16_be = bytes(
+        int("".join(hex_pair), 16) for hex_pair in hex_bytes
+    )
+    return emoji_utf_16_be.decode("utf-16-be")
+
+
+def emoji_to_noto_name(emoji):
```
Comment (Mic92) [inline pkgs/applications/networking/instant-messengers/signal-desktop/copy-noto-emoji.py:39]:
```suggestion
def emoji_to_noto_name(emoji: str) -> str:
```

**Comment by Mic92** [comment]
Looks good to me apart from my typing nitpicks above.

**Comment by teutat3s** [comment]
Result of `nixpkgs-review pr 337161` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>signal-desktop</li>
    <li>signal-desktop-beta</li>
  </ul>
</details>

**Review by teutat3s — APPROVED** [review]
Tested working by starting `signal-desktop` and testing a few emoji.

**Comment by teutat3s** [comment]
Nit: adding yourself as maintainer should be a separate commit.

**Comment by emilazy** [comment]
Nit addressed.

**Comment by stephen-huan** [comment]
For anyone wondering why the emojis changed and stumbled across this PR, an [overlay](https://github.com/stephen-huan/nixos-config/commit/c1af29d21687f8237df2ec8cff263b6344f83e86) to revert the change is

```nix
final: prev:

let
  self = prev.signal-desktop;
in
{
  # revert https://github.com/NixOS/nixpkgs/pull/337161
  signal-desktop = self.overrideAttrs (previousAttrs: {
    src = final.fetchurl {
      inherit (previousAttrs.src) url;
      hash = "";
    };

    nativeBuildInputs = previousAttrs.nativeBuildInputs ++ [ final.dpkg ];

    unpackPhase = ''
      runHook preUnpack

      dpkg-deb -x $src .

      runHook postUnpack
    '';

    installPhase =
      let
        lines = final.lib.splitString "\n" previousAttrs.installPhase;
        line =
          "# Copy the Noto Color Emoji PNGs into the ASAR contents. See `src`";
        index = final.lib.lists.findFirstIndex (x: x == line) null lines;
      in
      builtins.concatStringsSep "\n"
        ((final.lib.take index lines) ++ [ "runHook postInstall" ]);
  });
}
```

> I considered offering a `withUnfree` flag to allow the use of the bundled and downloaded Apple emoji, but it meaningfully complicates the derivation. If people feel very strongly about keeping the unlicensed Apple emoji, I can rework this change to add a flag at the cost of additional code complexity. (Although it’s arguably more `lib.licenses.illegal` than `lib.licenses.unfree`…)

I can only speak for myself, but I would prefer a flag. I understand why nixpkgs doesn't want to take risks here, so I don't feel strongly about adding the flag (and letting individual users overlay if they so prefer).

**Comment by emilazy** [comment]
If you want to open a PR to add a flag for it I’m happy to review as long as it defaults to `false` and sets the `license` appropriately. (Personally I think the overly‐compressed copies of the Apple emoji they ship look pretty awful anyway, and the privacy hit from downloading the jumbomoji packs on the fly is probably inconsequential but a little uncomfortable anyway, but preferences are preferences.)

**Pull request merged**